File: Interview/merge-k-sorted-lists.java
//  Merge K Sorted Lists

// Time O(NlogK), Space O(N)
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        List<ListNode> t = new ArrayList<ListNode>();
        for (ListNode x : lists)
            t.add(x);
        if (t.size() <= 0)
            return null;
        while (t.size() > 1) {
            int size = t.size();
            for (int i=0; i<size/2; i++) {
                ListNode node = mergeLists(t.get(0), t.get(1));
                t.remove(0); t.remove(0);
                t.add(node);
            }
        }
        return t.get(0);
    }
    
    private ListNode mergeLists(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(-1);
        ListNode p1 = l1, p2 = l2, p = dummy;
        while (p1!=null && p2!=null) {
            if (p1.val < p2.val) {
                p.next = new ListNode(p1.val);
                p1 = p1.next;
            }
            else {
                p.next = new ListNode(p2.val);
                p2 = p2.next;
            }
            p = p.next;
        }
        while (p1 != null) {
            p.next = new ListNode(p1.val);
            p1 = p1.next;
            p = p.next;
        }
        while (p2 != null) {
            p.next = new ListNode(p2.val);
            p2 = p2.next;
            p = p.next;
        }
        return dummy.next;
    }
}

File: leetcode/merge-k-sorted-lists.cc
//  Merge K Sorted Lists

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 0)  return nullptr;
        if (lists.size() == 1)  return lists[0];
        while (lists.size() > 1) {
            ListNode *l1 = lists[0], *l2 = lists[1];
            lists.erase(lists.begin());
            lists.erase(lists.begin());
            lists.push_back(mergeSortedLists(l1, l2));
        }
        return lists[0];
    }
    
    ListNode *mergeSortedLists(ListNode *l1, ListNode *l2) {
        if (!l1 || !l2)
            return l1 ? l1 : l2;
        ListNode *dummy = new ListNode(-1);
        ListNode *p1 = l1, *p2 = l2, *p = dummy;
        ListNode *temp;
        while (p1 && p2) {
            if (p1->val < p2->val) {
                temp = p1->next;
                p->next = p1;
                p1 = temp;
            }
            else {
                temp = p2->next;
                p->next = p2;
                p2 = temp;
            }
            p = p->next;
        }
        p->next = p1 ? p1 : p2;
        return dummy->next;
    }
};